source/common/conn_pool/conn_pool_base.cc

```diff
@@ -608,11 +660,27 @@ void ConnPoolImplBase::onPendingStreamCancel(PendingStream& stream,
     state_.decrPendingStreams(1);
     stream.removeFromList(pending_streams_);
   }
-  if (policy == Envoy::ConnectionPool::CancelPolicy::CloseExcess && !connecting_clients_.empty() &&
-      connectingConnectionIsExcess()) {
-    auto& client = *connecting_clients_.front();
-    transitionActiveClientState(client, ActiveClient::State::Draining);
-    client.close();
+  if (policy == Envoy::ConnectionPool::CancelPolicy::CloseExcess) {
+    if (!connecting_clients_.empty() &&
+        connectingConnectionIsExcess(*connecting_clients_.front())) {
+      auto& client = *connecting_clients_.front();
+      transitionActiveClientState(client, ActiveClient::State::Draining);
+      client.close();
+    } else if (!early_data_clients_.empty()) {
+      ActiveClient* idle_client{nullptr};
+      for (ActiveClientPtr& client : early_data_clients_) {
+        if (client->numActiveStreams() == 0) {
+          if (connectingConnectionIsExcess(*client)) {
+            idle_client = client.get();
+          }
+          break;
+        }
+      }
+      if (idle_client != nullptr) {
+        transitionActiveClientState(*idle_client, ActiveClient::State::Draining);
+        idle_client->close();
+      }
+    }
   }
 
   host_->cluster().stats().upstream_rq_cancelled_.inc();
```

Comment:
If we hit the "break" on line 676, should this stat be incremented?

Reply:
This is outside of `if (policy == Envoy::ConnectionPool::CancelPolicy::CloseExcess)` block, so it is always incremented.